pkg/skaffold/runner/generate_pipeline.go

```diff
@@ -0,0 +1,388 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runner
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+
+	tekton "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	yamlv2 "gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	// for testing
+	inputReader = bufio.NewReader(os.Stdin)
+)
+
+type StringReader interface {
+	ReadString(delim byte) (string, error)
+}
+
+func (r *SkaffoldRunner) GeneratePipeline(ctx context.Context, out io.Writer, config *latest.SkaffoldConfig, fileOut string) error {
+	err := createSkaffoldProfile(config)
+	if err != nil {
+		return errors.Wrap(err, "setting up profile")
+	}
+
+	fmt.Println("Generating Pipeline...")
+
+	// Generate git resource for pipeline
+	gitResource, err := generateGitResource()
+	if err != nil {
+		return errors.Wrap(err, "generating git resource for pipeline")
+	}
+
+	// Generate build task for pipeline
+	var tasks []*tekton.Task
+	taskBuild, err := generateBuildTask(config.Pipeline.Build)
+	if err != nil {
+		return errors.Wrap(err, "generating build task")
+	}
+	tasks = append(tasks, taskBuild)
+
+	// Generate deploy task for pipeline
+	taskDeploy, err := generateDeployTask(config.Pipeline.Deploy)
+	if err != nil {
+		return errors.Wrap(err, "generating deploy task")
+	}
+	tasks = append(tasks, taskDeploy)
+
+	// Generate pipeline from git resource and tasks
+	pipeline, err := generatePipeline(tasks)
+	if err != nil {
+		return errors.Wrap(err, "generating tekton pipeline")
+	}
+
+	// json.Marshal all pieces of pipeline, then convert all jsons to yamls
+	var jsons [][]byte
+	bGitResource, err := json.Marshal(gitResource)
+	if err != nil {
+		return errors.Wrap(err, "marshaling git resource")
+	}
+	jsons = append(jsons, bGitResource)
+	for _, task := range tasks {
+		bTask, err := json.Marshal(task)
+		if err != nil {
+			return errors.Wrap(err, "marshaling task")
+		}
+		jsons = append(jsons, bTask)
+	}
+	bPipeline, err := json.Marshal(pipeline)
+	if err != nil {
+		return errors.Wrap(err, "marshaling pipeline")
+	}
+	jsons = append(jsons, bPipeline)
+
+	var output bytes.Buffer
+	for _, item := range jsons {
+		itemYaml, err := yaml.JSONToYAML(item)
+		if err != nil {
+			return errors.Wrap(err, "converting jsons to yamls")
+		}
+		output.Write(append(itemYaml, []byte("---\n")...))
+	}
+
+	// write all yaml pieces to output
+	if err := ioutil.WriteFile(fileOut, output.Bytes(), 0755); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func generateGitResource() (*tekton.PipelineResource, error) {
```

Comment:
I think it would be useful to move all of these helper function into a pkg so the code is easier to find, like `pkg/skaffold/pipeline`

then you could have a file for each, like

`git_resource.go` and `git_resource_test.go` for the unit tests.